### PR TITLE
add serializes and deserializes tactical verbs to ontology

### LINKED ISSUE #269: serializes and deserializes tactical verbs
### Serialization

**Issue Overview:**

This issue proposes the addition of new tactical verbs related to serialization and deserialization in the ontology. Serialization and deserialization are fundamental processes in computing, particularly relevant for the storage, transmission, and reconstruction of data structures or object states. 

**Proposed Terms and Definitions:**

1. **:serializes**
    - **Type:** `owl:ObjectProperty`
    - **Label:** "serializes"
    - **SubPropertyOf:** `:associated-with, :hardens, :produces`
    - **DefinedBy:** `<http://dbpedia.org/resource/Serialization>`
    - **Definition:** "x serializes y: A subject x converts an artifact y into a serialized form. This process involves translating a data structure or object state into a format that can be stored or transmitted and reconstructed later."

2. **:deserializes**
    - **Type:** `owl:ObjectProperty`
    - **Label:** "deserializes"
    - **SubPropertyOf:** `:associated-with, :produces`
    - **DefinedBy:** `<http://dbpedia.org/resource/Deserialization>`
    - **Definition:** "x deserializes y: A subject x converts an artifact y from a serialized form back into its original form. This process involves extracting a data structure from a series of bytes to recreate a semantically identical clone of the original object."

3. **:DeserializationFunction**
    - **Type:** `owl:Class`
    - **Label:** "Deserialization Function"
    - **SubClassOf:** `:Subroutine`
    - **Restriction:** `[ owl:onProperty :deserializes ; owl:someValuesFrom :DigitalArtifact ]`
    - **Definition:** "A function that takes serialized data as input and deserializes it, typically using data parsing methods to reconstruct the original data structure or object."

4. **:SerializationFunction**
    - **Type:** `owl:Class`
    - **Label:** "Serialization Function"
    - **SubClassOf:** `:Subroutine`
    - **Restriction:** `[ owl:onProperty :serializes ; owl:someValuesFrom :DigitalArtifact ]`
    - **Definition:** "A function that serializes data by translating a data structure or object state into a format that can be stored or transmitted and later reconstructed."

5. **:encrypts**
    - **Type:** `owl:ObjectProperty`
    - **Label:** "encrypts"
    - **SubPropertyOf:** `:serializes, :hardens, :associated-with`
    - **DefinedBy:** `<http://wordnet-rdf.princeton.edu/id/00996121-v>`
    - **Definition:** "x encrypts y: The entity x converts the ordinary representation of a digital artifact y into a secret code, protecting its confidentiality."

6. **:decrypts**
    - **Type:** `owl:ObjectProperty`
    - **Label:** "decrypts"
    - **SubPropertyOf:** `:deserializes`
    - **DefinedBy:** `<http://wordnet-rdf.princeton.edu/id/00996499-v>`
    - **Definition:** "x decrypts y: A subject x converts an artifact y from a serialized form back into its original form by reversing the encryption process, restoring its original state."

**Rationale:**

Serialization and deserialization are crucial processes in computing, enabling data storage, transmission, and reconstruction. I think it makes sense that :encrypts is a subproperty of :serializes and to add its complement of :decrypts as well. 

**References:**
1. [Serialization - Wikipedia](https://en.wikipedia.org/wiki/Serialization)
2. [WordNet Definitions](http://wordnetweb.princeton.edu/perl/webwn)

**Turtle Syntax:**

```turtle
@prefix : <http://d3fend.mitre.org/ontologies/d3fend.owl#> .
@prefix dcterms: <http://purl.org/dc/terms/> .
@prefix owl: <http://www.w3.org/2002/07/owl#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix skos: <http://www.w3.org/2004/02/skos/core#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

:serializes a owl:ObjectProperty ;
    rdfs:label "serializes" ;
    rdfs:subPropertyOf :associated-with,
        :hardens,
        :produces ;
    rdfs:isDefinedBy <http://dbpedia.org/resource/Serialization> ;
    :definition "x serializes y: A subject x converts an artifact y into a serialized form. This process involves translating a data structure or object state into a format that can be stored or transmitted and reconstructed later." .

:deserializes a owl:ObjectProperty ;
    rdfs:label "deserializes" ;
    rdfs:subPropertyOf :associated-with,
        :produces ;        
    rdfs:isDefinedBy <http://dbpedia.org/resource/Deserialization> ;    
    :definition "x deserializes y: A subject x converts an artifact y from a serialized form back into its original form. This process involves extracting a data structure from a series of bytes to recreate a semantically identical clone of the original object." .

:DeserializationFunction a owl:Class ;
    rdfs:label "Deserialization Function" ;
    rdfs:subClassOf :Subroutine,
        [ a owl:Restriction ;
            owl:onProperty :deserializes ;
            owl:someValuesFrom :DigitalArtifact ] ;
    :definition "A function that takes serialized data as input and deserializes it, typically using data parsing methods to reconstruct the original data structure or object." .

:SerializationFunction a owl:Class ;
    rdfs:label "Serialization Function" ;
    rdfs:subClassOf :Subroutine, 
        [ a owl:Restriction ;
            owl:onProperty :serializes ;
            owl:someValuesFrom :DigitalArtifact ] ;
    :definition "A function that serializes data by translating a data structure or object state into a format that can be stored or transmitted and later reconstructed." .

:encrypts a owl:ObjectProperty ;
    rdfs:label "encrypts" ;
    rdfs:subPropertyOf :serializes,
        :hardens,
        :associated-with ;
    rdfs:isDefinedBy <http://wordnet-rdf.princeton.edu/id/00996121-v> ;
    :definition "x encrypts y: The entity x converts the ordinary representation of a digital artifact y into a secret code, protecting its confidentiality." .

:decrypts a owl:ObjectProperty ;
    rdfs:label "decrypts" ;
    rdfs:subPropertyOf :deserializes ;
    rdfs:isDefinedBy <http://wordnet-rdf.princeton.edu/id/00996499-v> ;
    :definition "x decrypts y: A subject x converts an artifact y from a serialized form back into its original form by reversing the encryption process, restoring its original state." .
```

### PR DESCRIPTION
**Pull request by aamedina**
- Introduce :serializes and :deserializes
- Add restrictions to :SerializationFunction and :DeserializationFunction classes that uses the verbs
- assert :encrypts as a subproperty of :serializes
- define :decrypts as a subproperty of :deserializes

Please critique/review and I'll edit accordingly!

fixes #269

### CODE REVIEW

File: src/ontology/d3fend-protege.ttl [hunk 1]
```diff
@@ -330,7 +342,8 @@ Moving forward different distinctions of kinds of has-part (contains) relationsh
 :encrypts a owl:ObjectProperty ;
     rdfs:label "encrypts" ;
     rdfs:subPropertyOf :associated-with,
-        :hardens ;
+        :hardens,
+        :serializes ;
```
Comment (netfl0) [inline src/ontology/d3fend-protege.ttl:346]: I am 80% sure this is true, @hack-sentinel can you comment on this superproperty declaration.

**Review by netfl0 — CHANGES_REQUESTED** [review]
My only question is added as a comment/task to @hack-sentinel , otherwise this looks create to me.

We have issues with linking to wordnet in general since they've changed what they're hosting. We ought to create a wordnet fork, if the license permits it, and create a stable github pages site for it.

File: src/ontology/d3fend-protege.ttl (hunk 1 above)
Comment (aamedina) [reply src/ontology/d3fend-protege.ttl:346]: @hack-sentinel What do you think?

Comment (netfl0) [reply src/ontology/d3fend-protege.ttl:346]: Lets flatten the superproperties to avoid confusion.

**Pull request closed without merging**